src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,255 @@
+ *   - visitor - call a visitor on each node we encounter
+ *   - Iterator/filterator based: FlowNode-by-FlowNode walking, after setup with head/blacklist
+ *
+ *  All operations can start with one or more "head" FlowNodes, and walk back from there, stopping when we hit blackList nodes
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
+
+    protected FlowNode _current;
+
+    protected FlowNode _next;
+
+    protected Collection<FlowNode> _blackList = Collections.EMPTY_SET;
+
+    /** Convert stop nodes to a collection that can efficiently be checked for membership, handling null if needed */
+    @Nonnull
+    protected Collection<FlowNode> convertToFastCheckable(@CheckForNull Collection<FlowNode> nodeCollection) {
+        if (nodeCollection == null || nodeCollection.size()==0) {
+            return Collections.EMPTY_SET;
+        } else  if (nodeCollection.size() == 1) {
+            return Collections.singleton(nodeCollection.iterator().next());
+        } else if (nodeCollection instanceof Set) {
+            return nodeCollection;
+        }
+        return nodeCollection.size() > 5 ? new HashSet<FlowNode>(nodeCollection) : nodeCollection;
+    }
+
+    /**
+     * Set up for iteration/analysis on a graph of nodes, initializing the internal state
+     * @param heads The head nodes we start walking from (the most recently executed nodes,
+     *              i.e. FlowExecution.getCurrentHeads()
+     * @param blackList Nodes that we cannot visit or walk past (useful to limit scanning to only nodes after a specific point)
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull Collection<FlowNode> heads, @CheckForNull Collection<FlowNode> blackList) {
+        if (heads == null || heads.size() == 0) {
+            return false;
+        }
+        Collection<FlowNode> fastEndNodes = convertToFastCheckable(blackList);
+        HashSet<FlowNode> filteredHeads = new HashSet<FlowNode>(heads);
+        filteredHeads.removeAll(fastEndNodes);
+
+        if (filteredHeads.size() == 0) {
+            return false;
+        }
+
+        reset();
+        _blackList = fastEndNodes;
+        setHeads(filteredHeads);
+        return true;
+    }
+
+    /**
+     * Set up for iteration/analysis on a graph of nodes, initializing the internal state
+     * @param head The head FlowNode to start walking back from
+     * @param blackList Nodes that we cannot visit or walk past (useful to limit scanning to only nodes after a specific point)
+     *                  null or empty collection means none
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull FlowNode head, @CheckForNull Collection<FlowNode> blackList) {
+        if (head == null) {
+            return false;
```

Comment:
But which callers would actually need to actually do a null check here? Using `@CheckForNull` increases the cognitive load on an API user so it should be avoided except in cases where there is either a _need_ to distinguish something from nothing, or where a caller could have done the null check but there are going to be tons of calls and it would get very noisy.